openraft/src/engine.rs

```diff
@@ -0,0 +1,259 @@
+    /// - The node has to contain no logs at all and the vote is the minimal value. See: [Conditions for initialization](https://datafuselabs.github.io/openraft/cluster-formation.html#conditions-for-initialization)
+    ///
+    /// Appending the very first log is slightly different from appending log by a leader or follower.
+    /// This step is not confined by the consensus protocol and has to be dealt with differently.
+    #[tracing::instrument(level = "debug", skip(self, entries))]
+    pub(crate) fn initialize<Ent: RaftEntry<NID>>(&mut self, entries: &mut [Ent]) -> Result<(), InitializeError<NID>> {
+        let l = entries.len();
+        assert_eq!(1, l);
+
+        let entry = &mut entries[0];
+
+        self.check_initialize()?;
+
+        let log_id = self.next_log_id();
+        entry.set_log_id(&log_id);
+
+        tracing::debug!("append initialization log id: {}", log_id);
+
+        self.commands.push(Command::AppendInputEntries { range: 0..l });
+        self.metrics_flags.set_data_changed();
+
+        if let Some(m) = entry.get_membership() {
+            let em = EffectiveMembership::new(Some(*entry.get_log_id()), m.clone());
+            self.state.effective_membership = Arc::new(em);
+
+            self.commands.push(Command::UpdateMembership { membership: m.clone() });
+            self.metrics_flags.set_cluster_changed();
+        } else {
+            panic!("Initialization log has to be a membership config entry",);
+        }
+
+        // TODO: set target state. This should be done by Engine but currently it is not.
+
+        self.commands.push(Command::MoveInputCursorBy { n: l });
+
+        Ok(())
+    }
+
+    /// Update the state for a new log entry to append. Update effective membership if the payload contains
+    /// membership config.
+    ///
+    /// TODO(xp): There is no check and this method must be called by a leader.
+    /// TODO(xp): metrics flag needs to be dealt with.
+    /// TODO(xp): if vote indicates this node is not the leader, refuse append
+    #[tracing::instrument(level = "debug", skip(self, entries))]
+    pub(crate) fn leader_append_entries<Ent: RaftEntry<NID>>(&mut self, entries: &mut [Ent]) {
+        let l = entries.len();
+        if l == 0 {
+            return;
+        }
+
+        for entry in entries.iter_mut() {
+            let log_id = self.next_log_id();
+            entry.set_log_id(&log_id);
+
+            tracing::debug!("append log id: {}", log_id);
+        }
+
+        self.commands.push(Command::AppendInputEntries { range: 0..l });
+        self.metrics_flags.set_data_changed();
+
+        for entry in entries.iter_mut() {
+            // TODO: if previous membership is not committed, reject a new change-membership propose.
+            //       unless the new config does not change any members but only learners.
+            if let Some(m) = entry.get_membership() {
+                let em = EffectiveMembership::new(Some(*entry.get_log_id()), m.clone());
```

Comment:
may be this update membership can merge into one function?

line 210 and line 168 has the same action.

Reply:
Good idea